### PR TITLE
Golang Homework 7.1 -  JSON

### PR DESCRIPTION
**Pull request by stevenstr**
**Home work 7.1 - Task:**
● Web server handling any request to it’s address and returning JSON

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	p := &Response{
+		Host:       r.Host,
+		UserAgent:  r.UserAgent(),
+		RequestURI: r.RequestURI,
+		Headers:    r.Header,
+	}
+
+	fmt.Fprintln(w, p)
+
+	fmt.Fprintln(w)
```
Comment (wshaman) [inline main.go]: This should return to a client a response formatted as 
```
HTTP response headers

Valid JSON body.
```
As you don't have to write headers by your own, and server will handle it for you, please just drop a JSON as a response body.

**Review by wshaman — COMMENTED** [review]
Good!
Nice idea to use mux here

File: Gopkg.toml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Gopkg.toml example
```
Comment (imatsiash-ib) [inline Gopkg.toml]: Do we need package manager here? I don't see any third-party packages you import.

File: main.go (hunk 1 above)
Comment (stevenstr) [reply main.go]: Done!)

File: Gopkg.toml (hunk 2 above)
Comment (stevenstr) [reply Gopkg.toml]: Thanks! 
I understood you)

File: main.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			Accept    string
+			UserAgent string
+		}{
+			r.Header.Get("Accept"),
+			r.Header.Get("User-agent"),
+		},
+	}
+
+	//pack for Write(interface)
+	pp, _ := json.Marshal(&p)
```
Comment (imatsiash-ib) [inline main.go]: Please do not ignore an error here.
In case of error you should return it to the client and set the status code to an error.

Comment (stevenstr) [reply main.go]: Done:) Thanks!

Comment (stevenstr) [reply main.go]: I can return it using panic and log. What you prefer in this case?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+			r.Header.Get("User-agent"),
+		},
+	}
+
+	//pack for Write(interface)
+	pp, err := json.Marshal(&p)
+	if err != nil {
+		log.Fatal(err)
+		//or
+		//panic(err)
```
Comment (imatsiash-ib) [inline main.go]: Instead of panic you should return an error message to the client. 
Also do not forget to set a status code back to the client.

Comment (stevenstr) [reply main.go]: \\\\\Done/////

**Review by wshaman — APPROVED** [review]
yep!

**Review by burov — APPROVED** [review]
Well done

File: main.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	UserAgent  string   `json:"user_agent"`
+	RequestURI string   `json:"request_uri"`
+	Headers    struct { //is correctly accoding the tasks
+		Accept    string
+		UserAgent string
+	} `json:"headers"`
+}
+
+func jsonHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
```
Comment (burov) [inline main.go:30]: It's better to set it in success case, because in case of an error you return not a json, client can fail to handle response

**Review by burov — APPROVED** [review]
I am ok with this solution but minor improvement possible

**Review by imatsiash-ib — APPROVED** [review]

**Pull request closed without merging**